smarts/env/wrappers/rgb_image.py

```diff
@@ -0,0 +1,88 @@
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+
+from typing import Any, Dict
+
+import gym
+import numpy as np
+
+
+class RGBImage(gym.ObservationWrapper):
+    """Filters SMARTS environment observation and returns only top-down RGB
+    image as observation.
+
+    If the base env's observation space is frame stacked using the FrameStack
+    wrapper, then the returned observation will be a numpy array of stacked
+    images with shape (image_width, image_height, 3*num_stack). Here, num_stack
+    refers to the number of frames stacked in the base env's observation.
+    """
+
+    def __init__(self, env: gym.Env, num_stack: int):
+        """
+        Args:
+            env (gym.Env): SMARTS environment to be wrapped.
+            num_stack (int): Use 1 if base env's observation space is not stacked,
+                else use the number of stacked frames in base env's observation.
+        """
+        super().__init__(env)
+        agent_specs = env.agent_specs
+
+        for agent_id in agent_specs.keys():
+            assert agent_specs[agent_id].interface.rgb, (
+                f"To use RGBImage wrapper, enable RGB "
+                f"functionality in {agent_id}'s AgentInterface."
+            )
+
+        self._num_stack = num_stack
+        assert self._num_stack > 0
+        self.observation_space = gym.spaces.Dict(
+            {
+                agent_id: gym.spaces.Box(
+                    low=0,
+                    high=255,
+                    shape=(
+                        agent_specs[agent_id].interface.rgb.width,
+                        agent_specs[agent_id].interface.rgb.height,
+                        3 * self._num_stack,
+                    ),
+                    dtype=np.uint8,
+                )
+                for agent_id in agent_specs.keys()
+            }
+        )
+
+    def observation(self, obs: Dict[str, Any]) -> Dict[str, np.ndarray]:
+        wrapped_obs = {}
+        for agent_id, agent_obs in obs.items():
+            if self._num_stack == 1:
+                agent_obs = [agent_obs]
+
+            images = []
+            for agent_ob in agent_obs:
+                image = agent_ob.top_down_rgb.data
+                images.append(image.astype(np.uint8))
```

Comment:
As far as I can tell, `num_stack` is intended cap the number of used frames from the frame stack. I think this limit should be applied here.

Reply:
A user could have stacked the wrappers in either of the following two approaches.

Approach1:
`HiwayEnv (obs_space: None) -> FrameStack (obs_space: None) -> RGBImage (obs_space: {agent_id: Box(w,h,3*num_stack)})`

Approach2:
`HiwayEnv (obs_space: None) -> RGBImage (obs_space: {agent_id: Box(w,h,3*1)})`

Previously in Approach1, it was difficult for `RGBImage` wrapper to know the underlying env’s `obs_space` since it is `None`. Hence, the need to feed in `num_stack` argument explicitly to `RGBImage` wrapper so that `RGBImage` wrapper outputs the proper `obs_space`.

In Approach2, `num_stack` argument to `RGBImage` would simply be fed as 1 by the user.

Going forward, I have changed `HiWayEnv`’s observation space. Additionally, I have changed the `FrameStack` wrapper’s observation space to `obs_space: {agent_id: Tuple([gym.Space]*num_stack)}`. With this change, the correct `num_stack` parameter would be auto deduced by `RGBImage` wrapper and the users do not need to feed this value to `RGBImage` wrapper anymore.

The changes for `HiWayEnv` and `Framestack` are in pr #1121. The necessary changes for this `RGBImage` wrapper is not made yet as it is dependent on the pr #1121 being merged first.

Reply:
Sure, yes, I was commenting on the mismatch between the `FrameStack` and `RGBImage` like with the following case:

```python
fs=FrameStack(env, num_stack=7)
RGBImage(fs, num_stack=3)
```

I think it should fail as early and clearly as possible which is why I think it should be checked or capped here (behaviour documented) to make sure the output matches the observation space rather than taking as many frames as the `FrameStack` will give.

Reply:
Ensured user supplied `num_stack` matches true num stack value.